sklearn/metrics/cluster/tests/test_supervised.py

```diff
@@ -254,12 +256,25 @@ def test_int_overflow_mutual_info_fowlkes_mallows_score():
 
 
 def test_entropy():
-    ent = entropy([0, 0, 42.0])
-    assert_almost_equal(ent, 0.6365141, 5)
+    assert_almost_equal(entropy([0, 0, 42.0]), 0.6365141, 5)
     assert_almost_equal(entropy([]), 1)
     assert entropy([1, 1, 1, 1]) == 0
 
 
+@pytest.mark.parametrize(
+    "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
+)
+def test_entropy_array_api(array_namespace, device, dtype_name):
+    xp = _array_api_for_tests(array_namespace, device)
+    float_labels = xp.asarray(np.asarray([0, 0, 42.0], dtype=dtype_name), device=device)
+    empty_int32_labels = xp.asarray([], dtype=xp.int32, device=device)
+    int_labels = xp.asarray([1, 1, 1, 1], device=device)
+    with config_context(array_api_dispatch=True):
+        assert_almost_equal(entropy(float_labels), 0.6365141, 5)
```

Comment:
This line is raising the following exception:

<details>

```python
____________________________ test_entropy_array_api[torch-cuda-float32] ____________________________

x = tensor(0.6365, device='cuda:0', dtype=torch.float64)

    @array_function_dispatch(_is_type_dispatcher)
    def iscomplexobj(x):
        """
        Check for a complex type or an array of complex numbers.
    
        The type of the input is checked, not the value. Even if the input
        has an imaginary part equal to zero, `iscomplexobj` evaluates to True.
    
        Parameters
        ----------
        x : any
            The input can be of any type and shape.
    
        Returns
        -------
        iscomplexobj : bool
            The return value, True if `x` is of a complex type or has at least
            one complex element.
    
        See Also
        --------
        isrealobj, iscomplex
    
        Examples
        --------
        >>> np.iscomplexobj(1)
        False
        >>> np.iscomplexobj(1+0j)
        True
        >>> np.iscomplexobj([3, 1+0j, True])
        True
    
        """
        try:
            dtype = x.dtype
>           type_ = dtype.type
E           AttributeError: 'torch.dtype' object has no attribute 'type'
dtype      = torch.float64
x          = tensor(0.6365, device='cuda:0', dtype=torch.float64)

/usr/local/lib/python3.10/dist-packages/numpy/lib/type_check.py:337: AttributeError

During handling of the above exception, another exception occurred:

array_namespace = 'torch', device = 'cuda', dtype_name = 'float32'

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    def test_entropy_array_api(array_namespace, device, dtype_name):
        xp = _array_api_for_tests(array_namespace, device)
        float_labels = xp.asarray(np.asarray([0, 0, 42.0], dtype=dtype_name), device=device)
        empty_int32_labels = xp.asarray([], dtype=xp.int32, device=device)
        int_labels = xp.asarray([1, 1, 1, 1], device=device)
        with config_context(array_api_dispatch=True):
>           assert_almost_equal(entropy(float_labels), 0.6365141, 5)

array_namespace = 'torch'
device     = 'cuda'
dtype_name = 'float32'
empty_int32_labels = tensor([], device='cuda:0', dtype=torch.int32)
float_labels = tensor([ 0.,  0., 42.], device='cuda:0')
int_labels = tensor([1, 1, 1, 1], device='cuda:0')
xp         = <module 'array_api_compat.torch' from '/usr/local/lib/python3.10/dist-packages/array_api_compat/torch/__init__.py'>

sklearn/metrics/cluster/tests/test_supervised.py:273:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
        args       = (tensor(0.6365, device='cuda:0', dtype=torch.float64), 0.6365141, 5)
        func       = <function assert_almost_equal at 0x7add6bcea830>
        kwds       = {}
        self       = <contextlib._GeneratorContextManager object at 0x7add6bcef4f0>
/usr/local/lib/python3.10/dist-packages/numpy/lib/type_check.py:339: in iscomplexobj
    type_ = asarray(x).dtype.type
        dtype      = torch.float64
        x          = tensor(0.6365, device='cuda:0', dtype=torch.float64)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = tensor(0.6365, device='cuda:0', dtype=torch.float64), dtype = None

    def __array__(self, dtype=None):
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        if dtype is None:
>           return self.numpy()
E           TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

dtype      = None
self       = tensor(0.6365, device='cuda:0', dtype=torch.float64)

/usr/local/lib/python3.10/dist-packages/torch/_tensor.py:1087: TypeError
```

</details>

when launching the tests with a cuda device on Google Colab with the help of this notebook: https://gist.github.com/EdAbati/ff3bdc06bafeb92452b3740686cc8d7c.

Since the entropy function always return a float Python scalar according to its docstring, let's just use the `float()` constructor before returning.